cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CreateExecutableTask.java

```diff
@@ -111,77 +90,88 @@ public void execute(BuildContext buildContext) {
+                outStream.putArchiveEntry(e);
                 outStream.write(service.toString().getBytes(StandardCharsets.UTF_8));
-                outStream.closeEntry();
+                outStream.closeArchiveEntry();
             }
-            // Copy dependency jar
-            // Copy dependency libraries
-            // Executable is created at give location.
-            // If no entry point is found we do nothing.
-        } catch (IOException | NullPointerException | IllegalAccessException e) {
+        } catch (IOException | NullPointerException e) {
             throw createLauncherException("unable to create the executable: " + e.getMessage());
         }
     }
 
     /**
-     * Copy jar file to executable fat jar.
+     * Copies a given jar file into the executable fat jar.
      *
-     * @param outStream     Executable jar out stream
-     * @param sourceJarFile Source file
-     * @param entries       Entries set wiil be used ignore duplicate files
+     * @param outStream     Output stream of the final uber jar.
+     * @param sourceJarFile Path of the source jar file.
+     * @param entries       Entries set will be used to ignore duplicate files.
      * @param services      Services will be used to temporary hold merged spi files.
-     * @throws IOException If file copy failed ioexception will be thrown
+     * @throws IOException If jar file copying is failed.
      */
-    private void copyJarToJar(ZipOutputStream outStream, String sourceJarFile, HashSet<String> entries,
+    private void copyJarToJar(ZipArchiveOutputStream outStream, String sourceJarFile, HashSet<String> entries,
                               HashMap<String, StringBuilder> services) throws IOException {
 
-        byte[] buffer = new byte[1024];
-        int len;
-        try (ZipInputStream inStream = new ZipInputStream(new FileInputStream(sourceJarFile))) {
-            for (ZipEntry e; (e = inStream.getNextEntry()) != null; ) {
-                String entryName = e.getName();
-                // Skip already copied files or excluded extensions.
-                if (e.isDirectory() || entries.contains(entryName) ||
-                        excludeExtensions.contains(entryName.substring(entryName.lastIndexOf(".") + 1))) {
-                    continue;
-                }
-                // SPIs will be merged first and then put into jar separately.
+        ZipFile zipFile = new ZipFile(sourceJarFile);
+        ZipArchiveEntryPredicate predicate = entry -> {
+            try {
+                String entryName = entry.getName();
                 if (entryName.startsWith("META-INF/services")) {
                     StringBuilder s = services.get(entryName);
                     if (s == null) {
                         s = new StringBuilder();
                         services.put(entryName, s);
                     }
                     char c = '\n';
+                    InputStream inStream = zipFile.getInputStream(entry);
```

Comment:
shall we wrap this with a BufferedInputStream as well?

Reply:
I have already tried wrapping input streams with the buffered ones, but couldn't had any noticeable gain. The reason might be we have to access input streams only for very few entries (only for the entries under `META-INF/services`). Anyway let have this included. :)